pkg/informer/informer.go

```diff
@@ -126,108 +158,150 @@ func NewDynamicDiscoverySharedInformerFactory(
-	if !ok {
-		return nil, fmt.Errorf("%T is not a CustomResourceDefinition", obj)
+// NewSelfDiscoveringDynamicSharedInformerFactory returns a factory for cluster-aware
+// shared informers that discovers new types and informs on updates to resources of
+// those types.
+// It retrieves GVR-related information from CRD informers.
+func NewSelfDiscoveringDynamicSharedInformerFactory(
+	cfg *rest.Config,
+	filterFunc func(obj interface{}) bool,
+	crdInformer kcpapiextensionsv1informers.CustomResourceDefinitionClusterInformer,
+	indexers cache.Indexers,
+) (*DiscoveringDynamicSharedInformerFactory, error) {
+	cfg = rest.AddUserAgent(rest.CopyConfig(cfg), "kcp-partial-metadata-informers")
+	metadataClusterClient, err := metadataclient.NewDynamicMetadataClusterClientForConfig(cfg)
+	if err != nil {
+		return nil, err
 	}
 
-	var ret []string
-
-	for _, v := range crd.Spec.Versions {
-		if !v.Served {
-			continue
-		}
-		ret = append(ret, byGroupVersionResourceKeyFunc(crd.Spec.Group, v.Name, crd.Spec.Names.Plural))
+	crdGVRSource, err := newCRDGVRSource(crdInformer.Informer())
+	if err != nil {
+		return nil, err
 	}
 
-	return ret, nil
+	return NewDiscoveringDynamicSharedInformerFactory(metadataClusterClient, filterFunc, nil, crdGVRSource, indexers)
+}
+
+// NewDiscoveringDynamicSharedInformerFactory returns a factory for cluster-aware
+// shared informers that discovers new types and informs on updates to resources of
+// those types.
+// It receives the GVR-related information by delegating to a GVRSource.
+func NewDiscoveringDynamicSharedInformerFactory(
+	dynamicClusterClient kcpdynamic.ClusterInterface,
+	filterFunc func(obj interface{}) bool,
+	tweakListOptions dynamicinformer.TweakListOptionsFunc,
+	gvrSource GVRSource,
+	indexers cache.Indexers,
+) (*DiscoveringDynamicSharedInformerFactory, error) {
+	f, err := NewGenericDiscoveringDynamicSharedInformerFactory[kcpcache.ScopeableSharedIndexInformer, kcpcache.GenericClusterLister](
+		func(gvr schema.GroupVersionResource, resyncPeriod time.Duration, indexers cache.Indexers) kcpinformers.GenericClusterInformer {
+			indexers[kcpcache.ClusterIndexName] = kcpcache.ClusterIndexFunc
+			indexers[kcpcache.ClusterAndNamespaceIndexName] = kcpcache.ClusterAndNamespaceIndexFunc
+			return kcpdynamicinformer.NewFilteredDynamicInformer(
+				dynamicClusterClient,
+				gvr,
+				resyncPeriod,
+				indexers,
+				nil,
+			)
+		},
+		filterFunc,
+		gvrSource,
+		indexers,
+	)
+
+	return &DiscoveringDynamicSharedInformerFactory{
+		GenericDiscoveringDynamicSharedInformerFactory: f,
+	}, err
+}
+
+// DiscoveringDynamicSharedInformerFactory is a factory for cluster-aware
+// shared informers that discovers new types and informs on updates to resources of
+// those types.
+// It offers an additional `Cluster()` method that returns a new shared informer factory
+// based on the main informer factory, but scoped for a given logical cluster.
+type DiscoveringDynamicSharedInformerFactory struct {
```

Comment:
We also don't need this type - just the one that it embeds.

Reply:
Where will you define the additional `Cluster()` method then ? It makes no sense for the cluster-unaware case not for the generic case ?

Reply:
Ah sorry, I was expecting the cluster-unaware type to be called Scopable and the aware type to be Scoped, and got confused by this type name.